File: src/lib/tauri/commands.ts
import { invoke } from '@tauri-apps/api/core';
import { open, save } from '@tauri-apps/plugin-dialog';
import type { ThemeSetting } from '../theme-resolve';
import type { EditorEngine } from '../stores.svelte';
import type { CommentThread } from '../comment-format';

export async function readFile(path: string): Promise<string> {
  return invoke<string>('read_file', { path });
}

export async function writeFile(path: string, content: string): Promise<void> {
  return invoke('write_file', { path, content });
}

export async function fileExists(path: string): Promise<boolean> {
  return invoke<boolean>('file_exists', { path });
}

/** Sets the Theme menu checkmarks; harmless no-op outside Tauri (browser dev). */
export function syncThemeMenu(preference: ThemeSetting): void {
  invoke('sync_theme_menu', { preference }).catch(() => {});
}

/** Sets the Editor Engine submenu checkmarks; harmless no-op outside Tauri. */
export function syncEngineMenu(engine: EditorEngine): void {
  invoke('sync_engine_menu', { engine }).catch(() => {});
}

/** Sets the "Include Live Render in Cmd+E" checkbox; harmless no-op outside Tauri. */
export function syncBetaInCycleMenu(enabled: boolean): void {
  invoke('sync_beta_in_cycle_menu', { enabled }).catch(() => {});
}

const FILE_FILTERS = [
  { name: 'All Supported', extensions: ['md', 'markdown', 'txt', 'csv', 'json', 'yml', 'yaml', 'toml', 'py', 'rs', 'ts', 'js', 'sh', 'env'] },
  { name: 'Markdown', extensions: ['md', 'markdown', 'txt'] },
  { name: 'Data', extensions: ['csv', 'json', 'yml', 'yaml', 'toml'] },
  { name: 'Code', extensions: ['py', 'rs', 'ts', 'js', 'sh'] },
  { name: 'All Files', extensions: ['*'] },
];

export async function showOpenDialog(): Promise<string | null> {
  const result = await open({
    multiple: false,
    filters: FILE_FILTERS,
  });
  return result as string | null;
}

export async function showSaveDialog(defaultName?: string): Promise<string | null> {
  const result = await save({
    defaultPath: defaultName,
    filters: FILE_FILTERS,
  });
  return result as string | null;
}

/** Matches `PendingOpen` in src-tauri/src/window.rs. */
export interface PendingOpen {
  path: string | null;
  content: string | null;
  cursor: number;
  topLine: number;
}

/**
 * Comment threads of a document, read from its `.mdmini_comments_<doc>.md`
 * sidecar. An absent sidecar is an empty list, not an error — most documents
 * have no comments, and the file only appears once the first one is written.
 */
export async function commentThreads(path: string): Promise<CommentThread[]> {
  return invoke<CommentThread[]>('comment_threads', { path });
}

/** Creates a thread anchored to `quote` and returns its new id. */
export async function commentCreate(
  path: string,
  line: number,
  quote: string,
  text: string
): Promise<string> {
  return invoke<string>('comment_create', { path, line, quote, text });
}

/**
 * Appends the user's own reply and returns the thread to `open`.
 *
 * The status matters: an agent's reply means "answered", but the user replying
 * again means they are waiting once more — and `open` is exactly what
 * `mdmini watch` emits an event for, so the agent gets woken by it.
 */
export async function commentReply(path: string, id: string, text: string): Promise<void> {
  return invoke('comment_reply', { path, id, text });
}

/** Marks a thread `resolved`. It stays in the file as history, never deleted. */
export async function commentResolve(path: string, id: string): Promise<void> {
  return invoke('comment_resolve', { path, id });
}

File: src-tauri/src/commands.rs
use std::fs;
use std::path::Path;
use tauri::command;

/// Returns and removes the pending payload for the calling window, if any.
/// Called by the frontend in onMount to pick up files passed via CLI args,
/// a new-window open, or a session restore.
#[command]
pub async fn get_pending_file(
    window: tauri::Window,
    state: tauri::State<'_, crate::window::PendingFiles>,
) -> Result<Option<crate::window::PendingOpen>, String> {
    let mut map = state.0.lock().map_err(|e| e.to_string())?;
    Ok(map.remove(window.label()))
}

#[command]
pub async fn read_file(path: String) -> Result<String, String> {
    fs::read_to_string(&path).map_err(|e| {
        if e.kind() == std::io::ErrorKind::InvalidData {
            "Cannot open: file is not valid text.".to_string()
        } else {
            format!("Failed to read file: {}", e)
        }
    })
}

#[command]
pub async fn write_file(path: String, content: String) -> Result<(), String> {
    let tmp_path = format!("{}.tmp", path);
    fs::write(&tmp_path, &content).map_err(|e| format!("Failed to write: {}", e))?;
    fs::rename(&tmp_path, &path).map_err(|e| {
        let _ = fs::remove_file(&tmp_path);
        format!("Failed to save: {}", e)
    })
}

#[command]
pub async fn file_exists(path: String) -> bool {
    Path::new(&path).exists()
}

/// Sets the Theme menu checkmarks to match the frontend's persisted
/// preference. Called on startup and on every theme change — the only
/// writer of these checkmarks (macOS toggles the clicked item natively;
/// this call corrects it).
#[command]
pub async fn sync_theme_menu(
    state: tauri::State<'_, crate::menu::ThemeMenuItems>,
    preference: String,
) -> Result<(), String> {
    state.sync(&preference);
    Ok(())
}

/// Sets the Editor Engine submenu checkmarks ("raw" | "live-preview" |
/// "live-render") to match the frontend's persisted engine. Same pattern as
/// `sync_theme_menu`: called on startup and on every engine change, since
/// macOS toggles the clicked item natively and this call corrects it.
#[command]
pub async fn sync_engine_menu(
    state: tauri::State<'_, crate::menu::EngineMenuItems>,
    engine: String,
) -> Result<(), String> {
    state.sync(&engine);
    Ok(())
}

/// Sets the "Include Live Render in Cmd+E" checkbox to match the frontend's
/// persisted `betaInCycle` flag. Independent of `sync_engine_menu` because
/// it isn't one of the three mutually exclusive engine choices.
#[command]
pub async fn sync_beta_in_cycle_menu(
    state: tauri::State<'_, crate::menu::EngineMenuItems>,
    enabled: bool,
) -> Result<(), String> {
    state.sync_beta_in_cycle(enabled);
    Ok(())
}

/// Comment threads of a document, read from its sidecar. A document with no
/// sidecar yet returns an empty list rather than an error — that is the normal
/// state for most files.
#[command]
pub async fn comment_threads(path: String) -> Result<Vec<crate::comments::Thread>, String> {
    crate::comments::load(std::path::Path::new(&path))
}

/// Creates a thread anchored to `quote` and returns its id.
///
/// The id is generated against the ids already in the file, so a hand-edited
/// file that happens to contain a colliding id cannot produce a duplicate.
#[command]
pub async fn comment_create(
    path: String,
    line: usize,
    quote: String,
    text: String,
) -> Result<String, String> {
    let doc = std::path::Path::new(&path);
    let taken: Vec<String> = crate::comments::load(doc)?
        .into_iter()
        .map(|thread| thread.id)
        .collect();
    let id = crate::comments::new_id_avoiding(doc, crate::comments::now_epoch(), &taken);
    crate::comments::append_thread(doc, &id, line, &quote, "You", &text)?;
    Ok(id)
}

/// Appends the user's own reply and puts the thread back to `open`.
///
/// `append_reply` sets `answered`, which is right for an agent but wrong here:
/// the user replying again means they are waiting once more, and `open` is
/// exactly what `mdmini watch` emits an event for — so this is what wakes the
/// agent for a follow-up question.
#[command]
pub async fn comment_reply(path: String, id: String, text: String) -> Result<(), String> {
    let doc = std::path::Path::new(&path);
    crate::comments::append_reply(doc, &id, "You", &text)?;
    crate::comments::set_status(doc, &id, crate::comments::Status::Open)
}

/// Marks a thread `resolved`. It stays in the file as history — threads are
/// never deleted, and pruning them is ordinary editing of a markdown file.
#[command]
pub async fn comment_resolve(path: String, id: String) -> Result<(), String> {
    crate::comments::set_status(
        std::path::Path::new(&path),
        &id,
        crate::comments::Status::Resolved,
    )
}
